### PR TITLE
Allow unauthenticated access to OFN API endpoints

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

I just retested the maps changes, and found the authentication on the new endpoint is really inconsistent. Most of the time it allows access when the user is logged out, but sometimes it denies the request with the error: "You must specify an API key". 

api_error

Logging in and logging out again seems to sometimes change the allow/deny behaviour from one state to the other, but I can't discern a pattern to it.

I've made a small change to the authentication:
- Allows API access without authentication, but only for OFN controllers which inherit from `BaseController`.
- Spree will now initialize `@current_api_user` with an empty `Spree::User` object (basically a "guest") if the current user is not logged in.
- This does not affect the authentication for existing Spree::Api endpoints.
- We should always use `authorize!` when required on API actions.

#### What should we test?

The new map page always load the modal when the user is logged out. Logging in and out doesn't change the outcome.

#### Release notes

Changed API authentication for custom endpoints

Changelog Category: Changed

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
@luisramos0 what do you think?

File: app/controllers/api/base_controller.rb [hunk 1]
```diff
@@ -21,5 +21,13 @@ def authenticate_user
       @current_api_user = try_spree_current_user
       super
     end
+
+    # Allows API access without authentication, but only for OFN controllers which inherit
+    # from BaseController. Actions that require authentication should all use #authorize!
```
Comment (luisramos0) [inline app/controllers/api/base_controller.rb]: maybe change to "from Api::BaseController (that inherits from Spree::Api::BaseController)"

File: app/controllers/api/enterprises_controller.rb [hunk 2]
```diff
@@ -77,5 +77,12 @@ def override_sells
     def override_visible
       params[:enterprise][:visible] = false
     end
+
+    # Allows API access without a logged in user for actions in this controller.
+    # Actions that require authentication should all use #authorize!
+    # @current_api_user will now initialize an empty Spree::User unless one is present.
```
Comment (sauloperez) [inline app/controllers/api/enterprises_controller.rb]: IMO it's worth explaining why we needed this and the tradeoff we made. That context will be invaluable the next time we end up here, don't you think?

Comment (kristinalim) [reply app/controllers/api/enterprises_controller.rb]: @sauloperez @Matt-Yorkley updated the comment - are you already happy with it?

**Comment by Matt-Yorkley** [comment]
I'm prepping a release, but I think we need to get this PR in, otherwise we'll have intermittent issues with the new map page...

**Comment by lin-d-hop** [comment]
Looks good to me as per Matt's testing notes.
Given the name of this PR I feel like I'm under testing this by just checking maps.

Reviewers: @kristinalim @sauloperez @luisramos0 just wanted to confirm that you agree this testing is sufficient?

https://docs.google.com/document/d/1pQCXLjf3-gRrJ8DAkEVeMrEXQc_rgmLT2pBhtAa7hr0/edit#

**Pull request merged**